### DISCUSSION TITLE
Configure private registries

### DISCUSSION THREAD

**Discussion by idoodler** [louislam/dockge#98] [Feature Request]
When I try to pull an image from my private GHCR it fails. The same stack works in Portainer where I setup my GHCP. 

An option to configure private registries would be nice 👍

**Comment by LorneCash** [comment]
I just need to authenticate with Docker Hub to avoid the rate limit.  I'm updating a few containers and I hit the limit for anonymous users very quickly.  I believe implementing this feature would also address my problem.

**Comment by jplegat** [comment]
+1

First, thanks a lot for creating Dockge 👍🏻!

Being able to integrate authentication - i.e.: GitHub with personal access tokens - would be great to manage Docker images in the container registry using the package namespace, straight from the Dockge UI.

`Error response from daemon: Head "https://ghcr.io/v2/username/packagename/manifests/image": unauthorized`

**Comment by canidam** [comment]
This is impossible to work without.. Any updates on this?
Even providing a WA such as injecting a config.json, or something sort of that would be great help.

**Reply by canidam** [reply]
I figured out a WA for the meantime.

Login to the container registry on the host. This will create ~/.docker/config.json file
Then mount this file to dockge container and restart. Now it can pull images from private repos.

A quick win can be to allow `docker login <repository> -u <username>` command from the terminal.

**Comment by Pekno** [comment]
Hello, any news on this ?

**Reply by muerwre** [reply]
This https://github.com/louislam/dockge/discussions/98#discussioncomment-11501670 works, should be done once for every new registry. 

I just click on "Shell" on dockge container inside dockge and exec `docker login https://your.registry.url`  in a shell inside dockge container.

**Comment by rickdgray** [comment]
OK here's how to make this work:

When Dockge runs the docker command, docker looks in the default location for a configuration, which is `/root/.docker/config.json`. So we need to mount a file to that location inside the Dockge container so that when you deploy, docker will automatically find it and authenticate.

Let's say your image is on a private ghcr. Here's what the config should look like:

```json
{
  "auths": {
    "ghcr.io": {
      "auth": "Z2l0aHViVXNlcjpnaXRodWJfdG9rZW4="
    }
  }
}
```

The auth should be a base 64 encoding of your github username and a personal access token. At time of writing, you have to use a classic PAT because the fine grained ones don't support packages yet. The token needs `repo` and `read:packages` privileges. You can get the encoded string like this:

```bash
echo -n 'GITHUB_USERNAME:PAT' | base64
```

Once you have your json, drop it wherever you like on your host. Personally I just put it adjacent to the `compose.yml` for Dockge. Wherever it is, you just need to reference it in said `compose.yml`. Here's what mine looks like:

```yaml
services:
  dockge:
    image: louislam/dockge:1
    restart: unless-stopped
    ports:
      - 5001:5001
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./data:/app/data
      - /opt/stacks:/opt/stacks
      - ./config.json:/root/.docker/config.json:ro
    environment:
      - DOCKGE_STACKS_DIR=/opt/stacks
```

After that, just run `docker compose up -d` again so that docker will recognize the change and redeploy Dockge. Now try to deploy your compose file inside Dockge and you should see that it is able to pull the image. 👌

**Reply by vinicioslc** [reply]
nice and clean aproach thanks
